### ISSUE TITLE
Water Line issue

### ISSUE THREAD

**Issue by rvanner** [TheGameCreators/GameGuruRepo#3096]
In this video;

https://youtu.be/ahhdUaD4KKw

Notice how the water line is between two lines? Should just be one.

The sound used when swimming isn't right. Let's check what FC5 does.

**Comment by ZakJudges** [comment]
@LeeBamberTGC @PaulSJ  I think this is caused by the water fog being applied below a fixed height, as if the water was a flat plane. Any ideas on an approach we could use to solve it?

**Comment by PaulSJ** [comment]
Yes it is caused by the underwater fog. The problem is that when the player is underwater there is nothing drawn between the player and the terrain underneath the water that can modify the render. 

For example when the player is above the water an looking down, the terrain is drawn first, then the water over the top of it, so the water can add the fog during its render. When the player is underwater the terrain is drawn first then the water is drawn above the player so it can't add fog to the terrain.

To resolve this, the underwater fog is applied to the terrain when the terrain is drawn, but it doesn't know exactly where the water line will be due to the waves. So it over estimates to make sure there are no gaps in teh fog. If the fog line was too low you'd get bright spots when the waves go above the fog line.

I think the only way to solve this fully would be to draw the underwater fog as a post process, after both the terrain and water have been drawn. So the terrain would no longer do anything special for under water, and the post process would add the fog to areas that were under water. This would only be needed when the player was under water.

It might need some way to detect the underside of the water, perhaps using the stencil buffer, not sure.

**Comment by ZakJudges** [comment]
@PaulSJ Thanks for the detailed explanation :) I saw there was some recent work done to the Wicked Engine for the underwater visuals: https://www.youtube.com/watch?v=EzUCu7te_8U
I don't know if we are still trying to get the most recent Wicked Changes in but maybe that would sort it?

**Comment by PaulSJ** [comment]
Looks like that would work, if you can figure out how it was done then we may be able to replicate it in our version. I suspect the water render was changed so that it renders some of the mesh in front of the camera, which would enable it to modify the scene. Either that or it uses a post process effect. RenderDoc might give you and idea of how it's working if you download a copy of the current WickedEngine and run the editor.
